### ISSUE TITLE
server blocks and can not get any http response

### ISSUE THREAD

**Issue by workingenius** [gin-gonic/gin#437]
hi everybody,
I'm new to golang and I'm working on our http api server with gin.
There are some time-consuming operation in some of our http handlers, and we just want the client to wait util results return so we didn't put those codes in another go routine or turn to another worker process.
While after running for hours, even the simplest api will not get any response.
e.g. such code would make clients wait till death:

``` go
api.GET("/_ping", func(c *gin.Context) {
    c.String(200, "pong")           
})
```

so I think this is not because of our time-consuming codes, but gin itself, right? Or are there any other possibilities I should check?
We copied gin source code to our proj several month ago with godep, so gin we are using differs from the latest.
thanks a lot

**Comment by javierprovecho** [comment]
Hi @workingenius,




Did you start the server? What kind of connection status did you get? The server should return an appropriate response, and before it, you can notify any go routine through a channel for example.




Another issue is that you are starting a blocking proccess before starting the server. I suggest to run either this process or gin server as a gorutine.




Give me more information so I can help you better.



NOTE: The information in this e-mail is confidential, being solely for the addressee named above. If you read this message and not the intended recipient, please note that it is completely forbidden any use, disclosure, distribution and / or duplication of this communication without the consent of the sender. If you have received this message in error, please notify us immediately by electronic mail and its elimination.

**Comment by workingenius** [comment]
@javierprovecho, 
thank you for your in time reply.
I'm sorry, my awful English may cause some misleading, let me explain some more.

There are 2 kinds of apis here, "for-write" and "for-read". "for-write" accept huge amount of data as input, in http chunked encoding;  the other, "for-read" accept simple json input while return huge chunked data. so both should last for a long time.
just like pseudocode below:
```go
        api.POST("/xxx", func(c *gin.Context) {   
            dec := decoder(c.Request.Body)
            for {
                var v [][]byte
                dec.Decode(&v)
                save_to_disk(v)
                // break if finish
            }
            c.Writer.WriteHeader(200)
            // then write some success info back
        })
```
you see, this api is just a time-consuming one, as the client keep feeding "records" in.
our main apis are like this, less than ten. Others are some traditional small apis that return at once.
the phenomenon is, the whole api refuse to process new requests. That's to say, already linked main apis still working, while new requests would wait and be blocked. Those small apis also block.

And it's not because of bad network. Other apis on the same machine work fine.

thank you
digging a reasonable explanation..

**Comment by nazwa** [comment]
This is a long shot but maybe it's the go scheduler that blocks.  Remember that go is concurrent, not parallel and the scheduler only uses certain types of operations to hand off the flow to other goroutines. You might simply be running out of cores to accept new requests. I had this problem before with busy loops.

Try putting `runtime.Gosched()` inside your long running loops to force goroutine switch.


```go
    api.POST("/xxx", func(c *gin.Context) {   
            dec := decoder(c.Request.Body)
            for {
                var v [][]byte
                dec.Decode(&v)
                save_to_disk(v)
                // break if finish

/// Allow another goroutine to run
/// ---->
runtime.Gosched()
/// <----
            }
            c.Writer.WriteHeader(200)
            // then write some success info back
        })
```

**Comment by javierprovecho** [comment]
@nazwa that makes sense! thank you for your answer.



NOTE: The information in this e-mail is confidential, being solely for the addressee named above. If you read this message and not the intended recipient, please note that it is completely forbidden any use, disclosure, distribution and / or duplication of this communication without the consent of the sender. If you have received this message in error, please notify us immediately by electronic mail and its elimination.

**Comment by workingenius** [comment]
Thank you for your answer. I guess it might be the exact problem, as I know little about golang concurrency mechanism. I'll reply after modifying my source code and check.

**Comment by nazwa** [comment]
Let me know how it goes :smile: Keep in mind that `gosched()` will slow down busy loops, so depending on your application you might want to only call it once every few iterations. But of course all of this depends on your application. 

If you are doing some very heavy lifting and lots of small requests in a single app, maybe you should think about splitting the app into two separate ones, so you can have better control over cpu load distribution. Assign each heavy worker to a single core and have a super lightweight frontend 'proxy' server that handles your small requests and proxies data to and from the heavy workers. This way your frontend app will never block the incoming requests.
